pkg/plugin/agent.go

```diff
@@ -0,0 +1,348 @@
+// SPDX-FileCopyrightText: Cristian Mogildea
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package plugin
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+)
+
+// ExecResponse struct is used to get the output of an executed command line
+type ExecResponse struct {
+	StdOut   string
+	StdErr   string
+	ExitCode int
+}
+
+const envDockerUser = "DOCKER_USER"
+const envDockerToken = "DOCKER_TOKEN"
+
+// ExecPlugin executes the plugin and returns nil if successful
+func ExecPlugin(cfg *Config) error {
+	resp, ctx, cli, err := PrepareContainer(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = StartPluginServer(cfg)
+	if err != nil {
+		return err
+	}
+
+	err = ExecAllPluginCmd(ctx, resp.ID, cfg)
+	if err != nil {
+		return err
+	}
+
+	err = StopContainer(resp.ID)
+	if err != nil {
+		return err
+	}
+
+	statusCh, errCh := cli.ContainerWait(ctx, resp.ID, container.WaitConditionNotRunning)
+	select {
+	case err := <-errCh:
+		if err != nil {
+			return err
+		}
+	case <-statusCh:
+	}
+
+	out, err := cli.ContainerLogs(ctx, resp.ID, types.ContainerLogsOptions{ShowStdout: true})
+	if err != nil {
+		return err
+	}
+
+	stdcopy.StdCopy(os.Stdout, os.Stderr, out)
+
+	return nil
+}
+
+// PrepareContainer pulls image from container registry and prepares container for execution
+func PrepareContainer(cfg *Config) (container.ContainerCreateCreatedBody, context.Context, *client.Client, error) {
```

Comment:
If `PrepareContainer` is only used in the `plugin` package it should be a private method.

Reply:
In general you should check which functions should really be public and which are only used within the package.

Reply:
Adjusted in 3c8a88c